components/BottomNavigation/src/MDCBottomNavigationBar.m

```diff
@@ -0,0 +1,354 @@
+              context:nil];
+    [item addObserver:self
+           forKeyPath:kMDCBottomNavigationBarBadgeValueString
+              options:NSKeyValueObservingOptionNew
+              context:nil];
+    [item addObserver:self
+           forKeyPath:kMDCBottomNavigationBarImageString
+              options:NSKeyValueObservingOptionNew
+              context:nil];
+    [item addObserver:self
+           forKeyPath:kMDCBottomNavigationBarTitleString
+              options:NSKeyValueObservingOptionNew
+              context:nil];
+  }
+}
+
+- (void)removeObserversFromTabBarItems {
+  for (UITabBarItem *item in self.items) {
+    @try {
+      [item removeObserver:self forKeyPath:kMDCBottomNavigationBarBadgeColorString];
+      [item removeObserver:self forKeyPath:kMDCBottomNavigationBarBadgeValueString];
+      [item removeObserver:self forKeyPath:kMDCBottomNavigationBarImageString];
+      [item removeObserver:self forKeyPath:kMDCBottomNavigationBarTitleString];
+    }
+    @catch (NSException *exception) {
+      if (exception) {
+        // No need to do anything if there are no observers.
+      }
+    }
+  }
+}
+
+- (void)removeBottomNavigationitemViews {
+  for (MDCBottomNavigationItemView *itemView in self.itemViews) {
+    [itemView removeFromSuperview];
+  }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
+                       context:(void *)context {
+  if (!context) {
+    NSInteger selectedItemNum = 0;
+    for (NSUInteger i = 0; i < self.items.count; i++) {
+      UITabBarItem *item = self.items[i];
+      if (object == item) {
+        selectedItemNum = i;
+        break;
+      }
+    }
+    MDCBottomNavigationItemView *itemView = _itemViews[selectedItemNum];
+    if ([keyPath isEqualToString:kMDCBottomNavigationBarBadgeColorString]) {
+      [itemView setBadgeColor:change[kMDCBottomNavigationBarNewString]];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationBarBadgeValueString]) {
+      [itemView setBadgeValue:change[kMDCBottomNavigationBarNewString]];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationBarImageString]) {
+      [itemView setImage:change[kMDCBottomNavigationBarNewString]];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationBarTitleString]) {
+      [itemView setTitle:change[kMDCBottomNavigationBarNewString]];
+    }
+  }
+}
+
+#pragma mark - Touch handlers
+
+- (void)didTouchDownButton:(UIButton *)button {
+  MDCBottomNavigationItemView *itemView = (MDCBottomNavigationItemView *)button.superview;
+  [itemView setCircleHighlightHidden:NO];
+}
+
+- (void)didTouchUpInsideButton:(UIButton *)button {
+  for (MDCBottomNavigationItemView *itemView in self.itemViews) {
+    if (itemView.button != button) {
+      [itemView setSelected:NO animated:YES];
+    }
+  }
+  MDCBottomNavigationItemView *itemView = (MDCBottomNavigationItemView *)button.superview;
```

Comment:
Might be simpler to bump the following two lines into the loop.

if (itemView.button == button) {
 // Newly selected item
 [itemView setSelected:YES animated:YES];
 itemView.circleHighlightHideen = YES;
} else {
  // De-select all other items
  [itemView setSelected:NO animated:YES]
}